### PR TITLE
Fix #7631:  16 out cargo support for industry directory

### LINKED ISSUE #7631: Assert when opening industry directory
## Version of OpenTTD

commit a83b80bacfa4aa5da2a6043136749b164ac2b191 (HEAD -> 7380)

Using dev version of FIRS (attached), _economy parameter set to "Steeltown"_.

## Expected result

OpenTTD doesn't assert when opening industry directory window.

## Actual result

OpenTTD asserts when opening industry directory window.

## Steps to reproduce

Add attached FIRS, set economy to "Steeltown".
Start a game.
Open industry directory window.
OpenTTD will assert.

ouch

[firs-1.tar.zip](https://github.com/OpenTTD/OpenTTD/files/3342311/firs-1.tar.zip)

### PR DESCRIPTION
**Pull request by glx22**
industry_directory

It's just a basic change. The window may need to be rewritten.
English.txt needs to be cleaned up

### CODE REVIEW

**Comment by James103** [comment]
Would it be possible to implement one of the following systems instead of just going to cargo sprites and hiding a bunch of information from players? This will retain more information in the industry directory window.

Plan A:
1-4 cargoes: Display the full cargo produced and transported ("100 tons of coal (50% transported)")
5-9 cargoes: Display only the cargo produced and the amount ("100 tons of coal, 200 tons of iron")
10+ cargoes: Display only the name of the cargo produced ("Coal, Iron, Gold, and Diamonds")

Plan B:
Cargoes 1-4: Display the full info regarding cargo produced and transported
Cargoes 5+: Collapse with a "...", "etc.", or "and more". Open industry window to view all cargoes there.
Example: "100 tons of coal, 200 tons of iron, 300 tons of gold, 400 tons of diamonds, and more"

**Comment by ldpl** [comment]
IMO this change makes industry window completely pointless as amount of produced cargo was the only useful info there. Also default cargo icons are terrible in this context. They're too small and not very recognizable.

**Comment by LordAro** [comment]
Why? Because currently the window hard crashes the game if your industry has more than 7 cargos produced/accepted.

I do agree that this isn't the correct solution though...

**Comment by andythenorth** [comment]
Could we do this:

* sort cargos by amount produced, highest > lowest
* take the first n (up to the limit of string params)
* append 'and x more...'

??

The rationale is that it's high production amounts that are looked for in this window.

**Comment by glx22** [comment]
So something like Unnamed, 15th Jan 1950 would be preferred.

**Comment by andythenorth** [comment]
Given that we have a hard-crash currently, I think that looks like a pretty good fix 👍 

Long-term, maybe we should change string handling so we can show all the cargos, but that also raises UI questions (horizontal overflow).  Let's fix the immediate crash.

**Review by michicc — DISMISSED** [review]

File: src/industry_gui.cpp [hunk 1]
```diff
@@ -1317,25 +1317,53 @@ class IndustryDirectoryWindow : public Window {
 		static CargoSuffix cargo_suffix[lengthof(i->produced_cargo)];
 		GetAllCargoSuffixes(CARGOSUFFIX_OUT, CST_DIR, i, i->type, indsp, i->produced_cargo, cargo_suffix);
 
-		/* Industry productions */
+		/* Get sorted industry productions */
+		struct Cargo {
```
Comment (LordAro) [inline src/industry_gui.cpp]: inline struct, that's a bit bold :)
"Cargo" is a bit generically named, are there no name conflicts/shadowing going on?

File: src/industry_gui.cpp [hunk 2]
```diff
@@ -1317,25 +1317,53 @@ class IndustryDirectoryWindow : public Window {
...
+			CargoID cargo;
+			uint16 production;
+			char *suffix;
+			uint transported;
+
+			/* Comparator to sort by descending production, then descending transported */
+			bool operator< (const Cargo &other) const
```
Comment (LordAro) [inline src/industry_gui.cpp]: this function appears to actually be implementing `>`... Confusing at best, incorrect at worst? Perhaps implement `>` (or `>=`) then define `<` in terms of that?

**Review by LordAro — CHANGES_REQUESTED** [review]

**Comment by glx22** [comment]
Oups I should have checked with mingw before pushing

**Comment by Eric-01** [comment]
While you are on it, what do you think about connecting your first presented idea with the existing one? To be more precise, I mean adding a cargo icon in front of each "x tonnes of y (z% transported)" string in both the corporate Industries window and all own windows of industries.

I think this shouldn't be hard to implement, and it could help you find the cargo type you're looking for, without the need to read each description.

**Review by LordAro — APPROVED** [review]

**Pull request merged**